### PR TITLE
Merge main into circuit

### PR DESCRIPTION
**Pull request by hdevalence**
This causes a test failure on the zero-length inner products (due to the changed logic in length checking for verification).

### CODE REVIEW

**Comment by hdevalence** [comment]
Updated to remove the 0-length IPP, which is only used for making constraint systems with no multipliers; those proofs could be better done using Schnorr proofs.

File: src/inner_product_proof.rs [hunk 1]
```diff
@@ -51,15 +51,6 @@ impl InnerProductProof {
...
-        if n == 0 {
-            return InnerProductProof {
-                L_vec: vec![],
-                R_vec: vec![],
-                a: Scalar::zero(),
-                b: Scalar::zero(),
-            };
-        }
-
         // All of the input vectors must have the same length.
```
Comment (oleganza) [inline src/inner_product_proof.rs]: Seems like `if n == 0` check got lost in this merge

Comment (oleganza) [reply src/inner_product_proof.rs]: Oh, I misunderstood your comment. I thought that's the check that you've added.

Why don't we make the API complete and allow zero multipliers, but improve the check in `verification_scalars` to handle zero correctly? Even if no-multiplier case is better handled by schnorr proofs, the user of a higher-level protocol may not even be exposed to the amount of multipliers and can accidentally create constraints without any.

Comment (hdevalence) [reply src/inner_product_proof.rs]: Any case where you would make a size-0 IPP you could just prove nothing, so I think that rejecting 0 is fine.

Comment (oleganza) [reply src/inner_product_proof.rs]: In a R1CS, internal 0-IPP would mean verifying linear constraints on a bunch of commitments, so it's implementing basically a schnorr proof, just with a few unnecessary challenges. So while IPP part proves nothing, CS proof is still useful.

Comment (oleganza) [reply src/inner_product_proof.rs]: I mean, 0-multiplier r1cs proof is proving a statement `<W_v, V> + W_c = 0` which is not nothing.

**Comment by hdevalence** [comment]
Would it be possible to not include the "R1CS" -> "ConstraintSystem" change in this PR? I didn't see it in #210, but I think that "R1CS" is a better name because it's more precise, it's a standard terminology, and it's shorter.

**Comment by cathieyun** [comment]
👀

**Pull request merged**